### PR TITLE
Make a plugin framework and move over trigger.

### PR DESCRIPTION
**Pull request by spxtr**
Huge PR, most of the lines are moving code but there's a fair bit of refactoring too. Sorry about that.

The framework works pretty nicely. You add a new package to `prow/plugins`, and add an `init` function, for example:

```
func init() {
	plugins.RegisterIssueCommentHandler("lgtm", handleIssueComment)
}

func handleIssueComment(pa *plugins.PluginAgent, ic github.IssueCommentEvent) error {
	// ...
}
```

Then when an issue comment event happens your function will be called if your plugin is enabled for the repo. It will be passed a `PluginAgent` which has clients to GitHub and Kubernetes and such, along with the event itself. We can later add caches to this if we need to save API tokens.

If you try to enable a plugin that doesn't exist, say you mistype it as `ltgm` in `plugins.yaml`, then you'll fail a unit test.

Still need to write some documentation, more comments, and more testing, but I think it's good for a first look.

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/test-infra/983)

### CODE REVIEW

File: prow/plugins/plugins.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+}
+
+type PluginAgent struct {
+	GitHubClient *github.Client
+	KubeClient   *kube.Client
+	JobAgent     *jobs.JobAgent
+
+	mut sync.Mutex
+	// Repo (eg "k/k") -> list of handler names.
+	ps map[string][]string
```
Comment (apelisse) [inline prow/plugins/plugins.go]: No idea what `ps` stands for

File: prow/plugins/plugins.go [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+// Load attempts to load config from the path. It returns an error if either
+// the file can't be read or it contains an unknown plugin.
+func (pa *PluginAgent) Load(path string) error {
+	pa.mut.Lock()
+	defer pa.mut.Unlock()
+	b, err := ioutil.ReadFile(path)
+	if err != nil {
+		return err
+	}
+	np := map[string][]string{}
```
Comment (apelisse) [inline prow/plugins/plugins.go]: OK, these variables names are randomly generated, it makes sense now :-D

File: prow/plugins/lgtm/lgtm.go [hunk 3]
```diff
@@ -21,25 +21,32 @@ import (
...
 	lgtmRe       = regexp.MustCompile(`(?mi)^\/lgtm\r?$`)
 	lgtmCancelRe = regexp.MustCompile(`(?mi)^\/lgtm cancel\r?$`)
 )
 
-type GitHubClient interface {
+func init() {
+	plugins.RegisterIssueCommentHandler(pluginName, handleIssueComment)
+}
+
+type gitHubClient interface {
```
Comment (apelisse) [inline prow/plugins/lgtm/lgtm.go]: why is this here?

Comment (apelisse) [inline prow/plugins/lgtm/lgtm.go]: You should probably use `githubClient` that does pretty much the same :D

File: prow/plugins/plugins.go [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+}
+
+type PullRequestHandler func(*PluginAgent, github.PullRequestEvent) error
+
+func RegisterPullRequestHandler(name string, fn PullRequestHandler) {
+	allPlugins[name] = struct{}{}
+	pullRequestHandlers[name] = fn
+}
+
+type PluginAgent struct {
```
Comment (apelisse) [inline prow/plugins/plugins.go]: I wish you didn't have to pass this entire type to each plugin. `GitHubClient` seems useful for most, it's much less true of `KubeClient` or `JobAgent`, not super critical.

File: prow/plugins/lgtm/lgtm.go (hunk 3 above)
Comment (spxtr) [inline prow/plugins/lgtm/lgtm.go]: It's so that I can define a minimal interface that is what I need to use for this one plugin. Should I choose to use a custom fake client for this plugin then I only need to implement those three methods.

The alternative is to have an interface in the `github` package that has every method, but then every fake needs to implement every method!

File: prow/plugins/plugins.go (hunk 2 above)
Comment (spxtr) [inline prow/plugins/plugins.go]: **n**ew **p**lugin.

I think Erick was a fan of my names in this package too :)

File: prow/plugins/plugins.go (hunk 1 above)
Comment (spxtr) [inline prow/plugins/plugins.go]: plugins

File: prow/plugins/plugins.go (hunk 4 above)
Comment (spxtr) [inline prow/plugins/plugins.go]: Somewhat agree, but I think it's reasonable. Each plugin should define their own interface for what functions they need out of the clients that are provided to them.

**Comment by krzyzacy** [comment]
you really need to implement `sudo /lgtm`
